app/utils/machine.ts

```diff
@@ -0,0 +1,91 @@
+import type { TransactionReceipt } from "@ethersproject/abstract-provider";
+import { assign, createMachine } from "xstate";
+
+export const createBlockchainTransactionStateMachine = <T>() => {
+  return createMachine(
+    {
+      predictableActionArguments: true, //recommended
+      id: "chain-write",
+      initial: "idle",
+      context: {
+        contractData: undefined,
+        transactionHash: undefined,
+        transactionReceipt: undefined,
+      },
+      schema: {
+        context: {} as { contractData?: T; transactionHash?: string; transactionReceipt?: TransactionReceipt },
+        events: {} as
+          | { type: "TRANSACTION_PREPARE" }
+          | { type: "TRANSACTION_READY"; data: T }
+          | { type: "TRANSACTION_CANCEL" }
+          | { type: "TRANSACTION_WRITE"; transactionHash: string }
+          | { type: "TRANSACTION_SUCCESS"; transactionReceipt: TransactionReceipt },
```

Comment:
I tend to create a type and import/use it here as it makes the machine smaller, same with context. Doesn't change anything though.
```
type TransactionEvents = 
   | { type: "TRANSACTION_PREPARE" }
   | { type: "TRANSACTION_READY"; data: T }
```
then 
events: {} as TransactionEvents